sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py

```diff
@@ -75,35 +76,53 @@ class CloudEvent(InternalCloudEvent, EventMixin):   #pylint:disable=too-many-ins
      unique for each distinct event.
     :type id: Optional[str]
     """
-
-    _validation = {
-        'source': {'required': True},
-        'type': {'required': True},
-    }
-
-    _attribute_map = {
-        'additional_properties': {'key': '', 'type': '{object}'},
-        'id': {'key': 'id', 'type': 'str'},
-        'source': {'key': 'source', 'type': 'str'},
-        'data': {'key': 'data', 'type': 'object'},
-        'data_base64': {'key': 'data_base64', 'type': 'bytearray'},
-        'type': {'key': 'type', 'type': 'str'},
-        'time': {'key': 'time', 'type': 'iso-8601'},
-        'specversion': {'key': 'specversion', 'type': 'str'},
-        'dataschema': {'key': 'dataschema', 'type': 'str'},
-        'datacontenttype': {'key': 'datacontenttype', 'type': 'str'},
-        'subject': {'key': 'subject', 'type': 'str'},
-    }
-
     def __init__(self, source, type, **kwargs):
         # type: (str, str, Any) -> None
-        kwargs.setdefault('id', uuid.uuid4())
-        kwargs.setdefault("source", source)
-        kwargs.setdefault("type", type)
-        kwargs.setdefault("time", dt.datetime.now(UTC()).isoformat())
-        kwargs.setdefault("specversion", "1.0")
-
-        super(CloudEvent, self).__init__(**kwargs)
+        self.source = source
+        self.type = type
+        self.specversion = kwargs.get("specversion", "1.0")
+        self.id = kwargs.get("id", str(uuid.uuid4()))
+        self.time = kwargs.get("time", dt.datetime.now(UTC()).isoformat())
+        self.data = kwargs.get("data", None)
+        self.datacontenttype = kwargs.get("datacontenttype", None)
+        self.dataschema = kwargs.get("dataschema", None)
+        self.subject = kwargs.get("subject", None)
+
+        for attr in kwargs:
+            setattr(self, attr, kwargs.get(attr))
+
+    @classmethod
+    def _from_generated(cls, generated, **kwargs):
+        if not generated:
+            return {}
+        return cls(
+            id=generated.id,
+            source=generated.source,
+            type=generated.type,
+            specversion=generated.specversion,
+            data=generated.data,
+            time=generated.time,
+            dataschema=generated.dataschema,
+            datacontenttype=generated.datacontenttype,
+            subject=generated.subject,
+            **kwargs
+        )
+
+    def _to_generated(self, **kwargs):
+        if isinstance(self.data, six.binary_type):
+            self.data = base64.b64encode(self.data)
+        return InternalCloudEvent(
+            id=self.id,
+            source=self.source,
+            type=self.type,
+            specversion=self.specversion,
+            data=self.data,
+            time=self.time,
+            dataschema=self.dataschema,
+            datacontenttype=self.datacontenttype,
+            subject=self.subject,
+            **kwargs
```

Comment:
Extensions should be passed part of additional_properties

```python
In [3]: c=CloudEvent(id=42, source="source", type="type", specversion="1.0", additional_properties={"foo":True})

In [4]: c.serialize()
Out[4]:
{'foo': True,
 'id': '42',
 'source': 'source',
 'type': 'type',
 'specversion': '1.0'}
```

Reply:
Apologies if I misremembered our call the other day but did we reach a conclusion about not using kwargs for extensions?  My fear being that we've just lost symmetry if someone **'s a serialized cloudevent back into itself.

Reply:
i took the liberty to slightly change the behavior to this 
```
CloudEvent(id=42, source="source", type="type", specversion="1.0", ext_as_kw = "hello",  extensions={"foo":True})
```

will serialize to 
```
{'foo': True,
'ext_as_kw ': 'hello',
 'id': '42',
 'source': 'source',
 'type': 'type',
 'specversion': '1.0'}
```
In other words, extensions can be accepted as both kwargs and an explicit extensions dictionary

Rationale is
1) This behavior is quite similar to the CloudEvent sdk's model which i think would benefit users jumping from there.
2) additional_properties as such can be sent by kwargs when possible 

I can disallow kwargs if there is a strong opinion against this

Reply:
The pitfall here is if someone has an additional property called additional_properties, we get "undefined behavior".  I seem to recall a mention in the call that we weren't TOO worried about that/could document it/reserved keywords and all that, but it's the one thing that irks me. 

This may be a good target for getting some broader user input on, having them hit this pitfall and see if it's dig-out-able or understandable to their eyes.

Reply:
> There should be one-- and preferably only one --obvious way to do it.
:)

What if you do `CloudEvent(foo=True, extensions={"foo":False})` for instance? I think we are opening a pandora's box of things we're unsure the day before the code complete. I would suggest we keep extensions but make this a priority question in user study.

Reply:
changed it to accepting only extensions - thank you both for the inputs

Reply:
Also @KieranBrantnerMagee , I'm catching up your comment, `additional_properties` is just for autorest model, I was explaining how to tell autorest model to serialize attributes that were not on the Swagger. It was not a comment about the external API of public CloudEvent for customer.